src/msgpack.light/MsgPackMemoryStreamReader.cs

```diff
@@ -19,29 +22,61 @@ public override byte ReadByte()
         {
             var temp = _stream.ReadByte();
             if (temp == -1)
+            {
                 throw ExceptionUtils.NotEnoughBytes(0, 1);
+            }
+
+            var result = (byte) temp;
+            _bytesGatheringBuffer?.Add(result);
 
-            return (byte)temp;
+            return result;
         }
 
         public override ArraySegment<byte> ReadBytes(uint length)
         {
-            var buffer = new byte[length];
-            var read = _stream.Read(buffer, 0, buffer.Length);
-            if (read < buffer.Length)
-                throw ExceptionUtils.NotEnoughBytes(read, buffer.Length);
+            var buffer = ReadBytesInternal(length);
+            _bytesGatheringBuffer?.AddRange(buffer);
```

Comment:
List of arrays maybe? To avoid lots of copying of internal arrays. In final method you can:
- create array of needed length
- copy only one time every array
- return byte[]